### DISCUSSION TITLE
ConvexHull query in Collision Query Demo results in AccessViolationException

### DISCUSSION THREAD

**Discussion by Cray-1** [bepu/bepuphysics2#220] [Q&A]
Hi, I tried to add a collision query with a convex hull. Seemed to work at first, but in some cases I get an AccessViolationException. I can reproduce this modifying the CollisionQueryDemo as follows:
1. Instead of the 5x5 box queries adding one (lines 224-238):
```
QuickList<Query> queries = new(1, BufferPool);
queries.Allocate(BufferPool) = new Query { Box = new Box(40, 15, 20), Pose = new RigidPose(new Vector3(0, 0.1f, 41), new Quaternion(-0.5f, -0.5f, -0.5f, 0.5f)) };
```
2. Increasing the size of queryWasTouched by one (line 241):
`BufferPool.Take(queries.Count + 1, out queryWasTouched);`
3. Adding a ConvexHull query (line 249):
```
QuickList<Vector3> points = new QuickList<Vector3>(8, BufferPool);
points.AllocateUnsafely() = new Vector3(0, 0, 0);
points.AllocateUnsafely() = new Vector3(0, 0, 2);
points.AllocateUnsafely() = new Vector3(2, 0, 0);
points.AllocateUnsafely() = new Vector3(2, 0, 2);
points.AllocateUnsafely() = new Vector3(0, 2, 0);
points.AllocateUnsafely() = new Vector3(0, 2, 2);
points.AllocateUnsafely() = new Vector3(2, 2, 0);
points.AllocateUnsafely() = new Vector3(2, 2, 2);
AddQueryToBatch(new ConvexHull(points, BufferPool, out _), new Vector3(), queries.Count, ref collisionBatcher);
```

Am I creating the convex hull wrong or what else could be the problem?

**Comment by RossNordby** [comment]
I'm amazed this has never been hit before- shapes that get cached by the caller for tasks that generate child tasks got inappropriately returned to the pool early, potentially resulting in memory stomping if a later task used that same memory region.

Should be fixed in 1c4d1c5571672517c473b1745cb7905adac7bbf1. Thanks for the report (again)!

**Comment by Cray-1** [comment]
Thanks for the quick reply (again) ;)
It indeed fixed the error of the reported example, but in a different case I get an AccessViolationException again. I don’t know if it is something similar or not and unfortunately the only way I could reproduce it in the Demo is by loading a static object from a file.
Therefore I added this simple function:
```
private Mesh ReadStl(string filepath, BufferPool bufferPool)
{
    BinaryReader reader = new(new FileStream(filepath, FileMode.Open, FileAccess.Read));
    reader.ReadBytes(80);
    int numTringles = reader.ReadInt32();
    bufferPool.Take<Triangle>(numTringles, out Buffer<Triangle> triangles);
    for (int triangleIndex = 0; triangleIndex < numTringles; triangleIndex++)
    {
        reader.ReadBytes(12);
        float x = reader.ReadSingle();
        float y = reader.ReadSingle();
        float z = reader.ReadSingle();
        Vector3 v1 = new(x, y, z);
        x = reader.ReadSingle();
        y = reader.ReadSingle();
        z = reader.ReadSingle();
        Vector3 v2 = new(x, y, z);
        x = reader.ReadSingle();
        y = reader.ReadSingle();
        z = reader.ReadSingle();
        Vector3 v3 = new(x, y, z);
        reader.ReadBytes(2);
        triangles[triangleIndex] = new Triangle(v1, v2, v3);
    }
    reader.Close();
    return new Mesh(triangles, new Vector3(1, 1, 1), bufferPool);
}
```
And loaded this mesh from the file as a static:
[MeshException.zip](https://github.com/bepu/bepuphysics2/files/8863656/MeshException.zip)
```
Mesh mesh = ReadStl("MeshException.stl", BufferPool);
Simulation.Statics.Add(new StaticDescription(new Vector3(), Simulation.Shapes.Add(mesh)));
```
Then this single box query results in the Exception during `collisionBatcher.Flush()`
```
queries.Allocate(BufferPool) = new Query { Box = new Box(105.905f, 193.368f, 126.943f), Pose = new RigidPose(new Vector3(300.3699f, 328.44067f, 430.81116f), new Quaternion(-0.1994829f, 0.1133757f, -0.35289928f, 0.9070913f)) };
```
It would be nice if you could look into this too.

**Reply by RossNordby** [reply]
Compound collision tests assume that there will be less than 262144 tests spawned by a single instance. The query overlaps more than that many triangles in the 350k triangle mesh, so that limit gets blown and bad stuff happens.

The bit packing is a little strained, so I'll probably first check if I can get away with widening the representation. Failing that, I can at least make the overflow not crash (though in that case, some child tests would be ignored).

**Reply by RossNordby** [reply]
New limit in aa6b0d8ae1e528bd715321fda43a15218b7d76a6 is 2^20, and exceeding it is no longer fatal. Extra child pairs are ignored.

*Could* go higher than this, though there are some tradeoffs and even a million child pairs in a single top level pair is not exactly ideal for performance.

Thanks again!
